semtypes/src/main/java/io/ballerina/types/typeops/ListProj.java

```diff
@@ -113,32 +118,30 @@ static SemType listProjPath(Context cx, SubtypeData k, Conjunction pos, Conjunct
                     Atom d = p.atom;
                     p = p.next;
                     lt = cx.listAtomType(d);
-                    TwoTuple intersected = listIntersectWith(members, rest,
-                            lt.members, lt.rest);
+                    TwoTuple intersected = listIntersectWith(cx.env, members, rest, lt.members(), lt.rest());
                     if (intersected == null) {
                         return NEVER;
                     }
                     members = (FixedLengthArray) intersected.item1;
-                    rest = (SemType) intersected.item2;
+                    rest = (CellSemType) intersected.item2;
                 }
             }
             if (fixedArrayAnyEmpty(cx, members)) {
                 return NEVER;
             }
             // Ensure that we can use isNever on rest in listInhabited
-            if (rest != NEVER && isEmpty(cx, rest)) {
-                rest = NEVER;
+            if (cellInnerVal(rest) != NEVER && isEmpty(cx, rest)) {
+                rest = cellContaining(cx.env, NEVER);
             }
         }
         // return listProjExclude(cx, k, members, rest, listConjunction(cx, neg));
         List<Integer> indices = ListOps.listSamples(cx, members, rest, neg);
-        int[] keyIndices;
-        TwoTuple projSamples = listProjSamples(indices, k);
-        TwoTuple sampleTypes = ListOps.listSampleTypes(cx, members, rest, indices);
-        return listProjExclude(cx, ((List<Integer>) projSamples.item1).toArray(new Integer[0]),
-                ((List<Integer>) projSamples.item2).toArray(new Integer[0]),
-                ((List<SemType>) sampleTypes.item1).toArray(new SemType[0]),
-                ((int) sampleTypes.item2), neg);
+        TwoTuple<List<Integer>, List<Integer>> projSamples = listProjSamples(indices, k);
+        TwoTuple<List<CellSemType>, Integer> sampleTypes = ListOps.listSampleTypes(cx, members, rest, indices);
+        return listProjExclude(cx, projSamples.item1.toArray(new Integer[0]),
```

Comment:
Since we are refactoring anyway, let's update to,
```suggestion
        return listProjExclude(cx, projSamples.item1.toArray(Integer::new),
```

Reply:
Fixed